### ISSUE TITLE
y axis scaling does not agree the fit report ? 

### ISSUE THREAD

**Issue by SunC0der** [Julian-Hochhaus/lmfit-additional-models#4]
Hi,
I am currently trying to fit 4 voigt functions to a spectrum with shirley background correction. However, during experimenting with the fit constrains I found that the actual values of the fit report do not seem to be reflected in the plots (see attached fit report and plot). It struck me when seeing

    "vg4_amplitude:  25.8909220 (init = 13)
    vg4_center:     610.000000 (init = 609.5)"

which is the brown curve in the plot, not even surpassing an intensity value of 5 imo. 

Any ideas why that could be ?

Thanks in advance

This is the code: 

```python

import sys
sys.path.append(
    "path/to/lmfit-additional-models"
)
from models import ShirleyBG
from lmfit.models import VoigtModel

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

def fit_four_voigt(
    directory,
    vg1_center,
    vg1_amplitude,
    vg2_center,
    vg2_amplitude,
    vg3_center,
    vg3_amplitude,
    vg4_center,
    vg4_amplitude,
):
    """Function to fit four voigt function with shirley background."""
    # General adaption:
    matplotlib.interactive(
        False
    )  # prevents spyder from floding the Ipython terminal with <Figure size 432x288 with 0 Axes> during batch processing

    # Load and label data:
    directory = Path(directory)

    fit_reports_directory = directory / "fit_reports"
    if fit_reports_directory.is_dir():
        pass
    else:
        fit_reports_directory.mkdir()

    fit_plots_directory = directory / "fit_plots"
    if fit_plots_directory.is_dir():
        pass
    else:
        fit_plots_directory.mkdir()

    data = pd.read_csv(
        directory / file.name,
        sep=",",
        header=None,
    )

    data.columns = ["Energy [eV]", "Intensity [a.u.]"]
    x = data.loc[:, "Energy [eV]"]
    y = data.loc[:, "Intensity [a.u.]"]

    # lmfit

    # Define models
    shirley_model = ShirleyBG(
        prefix="shirley_", independent_vars=["y"]
    )  # added independent_vars=['y'] due to the shirley needing a different independent variable than the default x

    linear_model = LinearModel(prefix="lin_", independent_vars=["x"])
    background_model = shirley_model  # + linear_model

    Pb2_4f_52_peak = VoigtModel(prefix="vg1_", independent_vars=["x"])
    Pb2_4f_72_peak = VoigtModel(prefix="vg2_", independent_vars=["x"])
    Pb0_4f_52_peak = VoigtModel(prefix="vg3_", independent_vars=["x"])
    Pb0_4f_72_peak = VoigtModel(prefix="vg4_", independent_vars=["x"])

    combined_model = (
        Pb2_4f_52_peak
        + Pb0_4f_52_peak
        + Pb2_4f_72_peak
        + Pb0_4f_72_peak
        + background_model
    )  # changed '-' to '+', because your data contain background + data, model needs to reflect that, background is included in the model and fit and not removed before fitting

    # Define parameters

    params = combined_model.make_params()  # lmfit.Parameters()

    params["vg3_amplitude"].set(
        value=50, min=40, max=60
    )  # Think about further constrains ! vg3_amplitude * 0.75

    params["vg3_center"].set(value=608)  # , min=607, max=609)
    params["vg3_sigma"].set(value=0.5)  # min=0)
    params["vg3_gamma"].set(value=0.5)

    params["vg4_amplitude"].set(
        value=13,
        min=10,
        )
    params["vg4_center"].set(value=609.5, min=609, max=610)
    params["vg4_sigma"].set(value=1.0)  # min=0)
    params["vg4_gamma"].set(value=1.0)

    params["vg1_amplitude"].set(
        expr="vg3_amplitude*0.75",
        min=0,  # max=50
    )
    params["vg1_center"].set(value=603)  # in=603, max=604)
    params["vg1_sigma"].set(expr="vg3_sigma")
    params["vg1_gamma"].set(expr="vg3_gamma")

    params["vg2_amplitude"].set(
        expr="vg4_amplitude*0.75",
        min=0,  # max=20
    )
    params["vg2_center"].set(expr="vg4_center - 4.4")
    params["vg2_sigma"].set(expr="vg4_sigma")
    params["vg2_gamma"].set(expr="vg4_gamma")

    # params.update(background_model.guess(y, x=x))
    params["shirley_const"].set(value=min(y))
    params["shirley_k"].set(value=0.003)

    # Fit
    result = combined_model.fit(
        y, params, x=x, y=y
    )  # <- The error is thrown #you need to pass all independent variables explicitly to the Model.fit() function

    # Display final results

    # print(result.fit_report())
    fit_report_name = Path("fit_report_" + file.name)
    with open(fit_report_name, "w") as f:
        f.write(result.fit_report())
    fit_report_name.rename(fit_reports_directory / fit_report_name)

    shirley = result.eval_components()["shirley_"]
    #lin = result.eval_components()["lin_"]

    # matplotlib
    fig, ax = plt.subplots()

    ax.scatter(x, y, s=1, c="black", label="data")
    # plt.plot(x, result.init_fit, "--", label="initial fit")
    ax.plot(x, result.best_fit, "-", label="best fit")  # -bg
    ax.plot(x, shirley, ".-", label="shirley bg")
    # ax.plot(x, lin, ".-", label="linear bg")

    comps = result.eval_components()
    ax.plot(x, comps["vg1_"], "--", label="voigt1")
    ax.plot(x, comps["vg2_"], "--", label="voigt2")
    ax.plot(x, comps["vg3_"], "--", label="voigt3")
    ax.plot(x, comps["vg4_"], "--", label="voigt4")

    ax.set_title("CTF3 fits with 4 voigt peak and shirley background")
    ax.set_ylabel("Intensity [a.u.]")
    ax.set_xlabel("E_kin [eV]")
    ax.legend()

    fig.savefig(str(fit_plots_directory / file.name) + ".png", dpi=300)
    plt.close()

# --------------------------------------------

start = time.time_ns()

data_path = Path(
    "/path/to/data/data_file.csv"
)

file_names = []

fit_four_voigt(data_path, 603, 30, 605, 100, 607, 50, 609, 15)


end = time.time_ns()

elapsed_time = end - start

print("Execution time:", elapsed_time, "nano seconds")

``` 
Plot and fit report: 
![test_spectrum csv](https://github.com/Julian-Hochhaus/lmfit-additional-models/assets/114517513/4491ba14-7152-4792-8c9f-0cd7c3ff5dfa)
[fit_report_test_spectrum.csv](https://github.com/Julian-Hochhaus/lmfit-additional-models/files/11566963/fit_report_test_spectrum.csv)

Data: 
[test_spectrum.csv](https://github.com/Julian-Hochhaus/lmfit-additional-models/files/11566965/test_spectrum.csv)

**Comment by Julian-Hochhaus** [comment]
I assume, the parameter names in lmfit are just confusing. The  amplitude is not the peak_height but the peak area.
As explained [in the other Issue](https://github.com/Julian-Hochhaus/lmfit-additional-models/issues/3#issuecomment-1563900581).
And quoting from lmfit docu:

> There are many peak-like models available. These include [GaussianModel](https://lmfit.github.io/lmfit-py/builtin_models.html#lmfit.models.GaussianModel), [LorentzianModel](https://lmfit.github.io/lmfit-py/builtin_models.html#lmfit.models.LorentzianModel), [VoigtModel](https://lmfit.github.io/lmfit-py/builtin_models.html#lmfit.models.VoigtModel), [PseudoVoigtModel](https://lmfit.github.io/lmfit-py/builtin_models.html#lmfit.models.PseudoVoigtModel), and some less commonly used variations. Most of these models are unit-normalized and share the same parameter names so that you can easily switch between models and interpret the results. **The amplitude parameter is the multiplicative factor for the unit-normalized peak lineshape, and so will represent the strength of that peak or the area under that curve.**

[Link](https://lmfit.github.io/lmfit-py/builtin_models.html)

**Comment by SunC0der** [comment]
Thank you, that naming "amplitude" didnt even make me think that it could mean something else then the height. I will try to fix the fits now and report how it went

**Comment by Julian-Hochhaus** [comment]
I am honestly not happy with lmfit's naming convention, because during fitting it's often much easier to guess the height in advance and use this guess for fast-converging fits. I don't get why they are not norming the model's maximum to 1 and multiplying by height and calculating the area afterward instead of calculating the height afterward, especially because it would roughly need the same computing amount and the area is not well-defined for all models.

**Comment by SunC0der** [comment]
I see why they would make the area a parameter, due to its meaning for symmetrical fits. But it would be good if one could - depending on the model - choose between using either of the three as parameter (FWHM, area or height).

I succeded in fitting with the change parameter and with basinhoppin as fitting method to fit the data - see plot below,

Thank you for your help !

![test_spectrum csv](https://github.com/Julian-Hochhaus/lmfit-additional-models/assets/114517513/e1022d63-688c-457b-9fc2-f607cad2bb98)

**Comment by Julian-Hochhaus** [comment]
@SunC0der Just a small tip, if you are interested in the brown component in the plot, you should probably measure several eVs more on the high kinetic energy side. At least the 2eV you measured on the left of the spectrum are necessary for reliable background calculation from the center of the rightmost peak to the right end of the spectrum.

**Comment by SunC0der** [comment]
@Julian-Hochhaus yes, the spectrum range is in my opinion not ideal either. However, there is a general problem with the spectrum regarding the analysis:  
1. zero value areas are on both high and lowenergy ends
2. sharp peaks seperate the zero_value areas from the actual data, which are not related to the sample afaik. I remove them before smoothing such that the smoothing doesnt get skewed. 

See plot below depicting three similar spectra in three colours. 

Do you have an idea how to better deal with these data ? Any advice, small or big welcome :-)

**Comment by Julian-Hochhaus** [comment]
What mode is your detector in during measurement? Is it static in "snapshot" mode or scanning over the energies (probably called withering mode)? I assume some kind of snapshot mode and that the weird signals you see at the edges are probably some kind of error caused by the detector. Maybe try measuring the same range energy range for an energy range where you do not expect signals at all or measure the same range without a sample in front of the detector.

I would remove these artifacts by setting upper and lower limit for the x-data. 
In addition, if I was you, I would give it a try without smoothing the data, propably that should work too.

**Comment by SunC0der** [comment]
I am unfortunately only at the receiving end of the experimental data and therefor do not know in which mode the data were taken. I will ask out of curiosity though and will try the fitting of unsmoothed data.
